File: 2061-painting-a-grid-with-three-different-colors/painting-a-grid-with-three-different-colors.go
const MOD = 1_000_000_007

func colorTheGrid(m int, n int) int {
    validStates := generateValidStates(m);
    transitions := buildTransitions(validStates, m);

    dp := make([]int, len(validStates));
    for i := range dp {
        dp[i] = 1;
    }

    for col := 1; col < n; col++ {
        newDp := make([]int, len(validStates));
        for i := range validStates {
            for _, prev := range transitions[i] {
                newDp[i] = (newDp[i] + dp[prev]) % MOD;
            }
        }
        dp = newDp;
    }

    result := 0;
    for _, count := range dp {
        result = (result + count) % MOD;
    }

    return result;
}

func generateValidStates(m int) [][]int {
    total := 1;
    for i := 0; i < m; i++ {
        total *= 3;
    }

    var validStates [][]int;
    for mask := 0; mask < total; mask++ {
        colors := decode(mask, m);
        if isValid(colors) {
            validStates = append(validStates, colors);
        }
    }

    return validStates;
}

func buildTransitions(validStates [][]int, m int) map[int][]int {
    transitions := make(map[int][]int);
    for i, a := range validStates {
        for j, b := range validStates {
            if canTransition(a, b) {
                transitions[i] = append(transitions[i], j);
            }
        }
    }

    return transitions;
}

func isValid(colors []int) bool {
    for i := 1; i < len(colors); i++ {
        if colors[i] == colors[i-1] {
            return false;
        }
    }

    return true;
}

func canTransition(a, b []int) bool {
    for i := 0; i < len(a); i++ {
        if a[i] == b[i] {
            return false;
        }
    }

    return true;
}

func decode(mask, m int) []int {
    colors := make([]int, m);
    for i := 0; i < m; i++ {
        colors[i] = mask % 3;
        mask /= 3;
    }
    
    return colors;
}

File: 2061-painting-a-grid-with-three-different-colors/painting-a-grid-with-three-different-colors.java
class Solution {
    private static final int MOD = 1_000_000_007;
    private List<int[]> validCols = new ArrayList<>();
    private Map<String, Integer> colIndex = new HashMap<>();
    private List<List<Integer>> transitions = new ArrayList<>();

    public int colorTheGrid(int m, int n) {
        generateValidColumns(m, new ArrayList<>());

        // map cấu hình thành index
        for (int i = 0; i < validCols.size(); i++) {
            String key = Arrays.toString(validCols.get(i));
            colIndex.put(key, i);
        }

        // tạo danh sách các transition
        for (int[] col1 : validCols) {
            List<Integer> list = new ArrayList<>();
            for (int i = 0; i < validCols.size(); i++) {
                int[] col2 = validCols.get(i);
                if (isCompatible(col1, col2)) {
                    list.add(i);
                }
            }
            transitions.add(list);
        }

        int k = validCols.size();
        int[] dp = new int[k];
        Arrays.fill(dp, 1);

        for (int step = 1; step < n; step++) {
            int[] newDp = new int[k];
            for (int i = 0; i < k; i++) {
                for (int j : transitions.get(i)) {
                    newDp[i] = (newDp[i] + dp[j]) % MOD;
                }
            }
            dp = newDp;
        }

        int result = 0;
        for (int val : dp) {
            result = (result + val) % MOD;
        }

        return result;
    }

    private void generateValidColumns(int m, List<Integer> col) {
        if (col.size() == m) {
            int[] arr = new int[m];
            for (int i = 0; i < m; i++) arr[i] = col.get(i);
            validCols.add(arr);
            return;
        }

        for (int color = 0; color < 3; color++) {
            if (!col.isEmpty() && col.get(col.size() - 1) == color) continue;
            col.add(color);
            generateValidColumns(m, col);
            col.remove(col.size() - 1);
        }
    }

    private boolean isCompatible(int[] a, int[] b) {
        for (int i = 0; i < a.length; i++) {
            if (a[i] == b[i]) return false;
        }
        
        return true;
    }
}